lib/stdlib/src/supervisor.erl

```diff
@@ -708,20 +716,54 @@ do_restart(Reason, Child, State) when ?is_permanent(Child) ->
     restart(Child, State);
 do_restart(normal, Child, State) ->
     NState = del_child(Child, State),
-    {ok, NState};
+    do_auto_shutdown(Child, NState);
 do_restart(shutdown, Child, State) ->
     NState = del_child(Child, State),
-    {ok, NState};
+    do_auto_shutdown(Child, NState);
 do_restart({shutdown, _Term}, Child, State) ->
     NState = del_child(Child, State),
-    {ok, NState};
+    do_auto_shutdown(Child, NState);
 do_restart(Reason, Child, State) when ?is_transient(Child) ->
     ?report_error(child_terminated, Reason, Child, State#state.name),
     restart(Child, State);
 do_restart(Reason, Child, State) when ?is_temporary(Child) ->
     ?report_error(child_terminated, Reason, Child, State#state.name),
     NState = del_child(Child, State),
-    {ok, NState}.
+    do_auto_shutdown(Child, NState).
+
+do_auto_shutdown(_Child, State=#state{auto_shutdown = never}) ->
+    {ok, State};
+do_auto_shutdown(Child, State) when not ?is_significant(Child)->
+    {ok, State};
+do_auto_shutdown(_Child, State=#state{auto_shutdown = any_significant}) ->
+    {shutdown, State};
```

Comment:
Is that right? You don't check whether the child is significant here. Wouldn't a non-significant child exiting result in the supervisor shutting down as well?

Reply:
The clause at line 736 matches when the child is _not_ significant, so anything that gets to the clause at line 738 must be a significant child.

Reply:
Right. For some reason I completely missed that line. Good then.